src/vdaf/mastic.rs

```diff
@@ -0,0 +1,727 @@
+    fn eq(&self, other: &MasticOutputShare<V>) -> bool {
+        self.result == other.result
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Vdaf for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type Measurement = (VidpfInput, T::Measurement);
+    type AggregateResult = T::AggregateResult;
+    type AggregationParam = MasticAggregationParam;
+    type PublicShare = MasticPublicShare<VidpfWeight<T::Field>>;
+    type InputShare = MasticInputShare<T::Field, SEED_SIZE>;
+    type OutputShare = MasticOutputShare<VidpfWeight<T::Field>>;
+    type AggregateShare = MasticOutputShare<VidpfWeight<T::Field>>;
+
+    fn algorithm_id(&self) -> u32 {
+        self.algorithm_id
+    }
+
+    fn num_aggregators(&self) -> usize {
+        2
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn shard_with_random(
+        &self,
+        measurement_label: &VidpfInput,
+        measurement_weight: &VidpfWeight<T::Field>,
+        nonce: &[u8; 16],
+        vidpf_keys: [VidpfKey; 2],
+        szk_random: [Seed<SEED_SIZE>; 2],
+        opt_random: Option<Seed<SEED_SIZE>>,
+    ) -> Result<(<Self as Vdaf>::PublicShare, Vec<<Self as Vdaf>::InputShare>), VdafError> {
+        // Compute the measurement shares for each aggregator by generating VIDPF
+        // keys for the measurement and evaluating each of them.
+        let public_share = self.vidpf().gen_with_keys(
+            &vidpf_keys,
+            measurement_label,
+            measurement_weight,
+            nonce,
+        )?;
+
+        let leader_measurement_share = self
+            .vidpf()
+            .eval(
+                &vidpf_keys[0],
+                &public_share,
+                &VidpfInput::from_bools(&[false]),
+                nonce,
+            )?
+            .share
+            + self
+                .vidpf()
+                .eval(
+                    &vidpf_keys[0],
+                    &public_share,
+                    &VidpfInput::from_bools(&[true]),
+                    nonce,
+                )?
+                .share;
+        let helper_measurement_share = self
+            .vidpf()
+            .eval(
+                &vidpf_keys[1],
+                &public_share,
+                &VidpfInput::from_bools(&[false]),
+                nonce,
+            )?
+            .share
+            + self
+                .vidpf()
+                .eval(
+                    &vidpf_keys[1],
+                    &public_share,
+                    &VidpfInput::from_bools(&[true]),
+                    nonce,
+                )?
+                .share;
+
+        let szk_proof_shares = self.szk.prove(
+            leader_measurement_share.as_slice(),
+            helper_measurement_share.as_slice(),
+            measurement_weight.as_slice(),
+            szk_random,
+            opt_random,
+            nonce,
+        )?;
```

Comment:
In general, clone may result in a memory allocation, so we should avoid cloning if we can. Here it's easy, as you just need to unpack the array output by `self.szk.prove()`.

We should do the same thing for `vidpf_keys`, for tidiness if nothing else. Those clones are unlikely to result in allocations.

```suggestion
        let [leader_szk_proof_share, helper_szk_proof_share] = self.szk.prove(
            leader_measurement_share.as_slice(),
            helper_measurement_share.as_slice(),
            measurement_weight.as_slice(),
            szk_random,
            opt_random,
            nonce,
        )?;
        let [leader_vidpf_key, helepr_vidpf_key] = vidpf_keys;
```